internal/restic/rewind_reader.go

```diff
@@ -0,0 +1,90 @@
+package restic
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/restic/restic/internal/errors"
+)
+
+// RewindReader allows resetting the Reader to the beginning of the data.
+type RewindReader interface {
+	io.Reader
+
+	// Rewind rewinds the reader so the same data can be read again from the
+	// start.
+	Rewind() error
+
+	// Length returns the number of bytes that can be read from the Reader
+	// after calling Rewind.
+	Length() int
```

Comment:
shouldn't this be int64? I don't think it's reasonable to assume all uploaded files will be smaller than 2G.

Reply:
Good point, I'll change it to `int64`.